src/serialization/interval_log/mod.rs

```diff
@@ -0,0 +1,596 @@
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub struct Tag<'a>(&'a str);
+
+impl<'a> Tag<'a> {
+    /// Create a new Tag.
+    ///
+    /// If a disallowed character is present, this will return `None`.
+    pub fn new(s: &'a str) -> Option<Tag<'a>> {
+        if s.chars()
+            .any(|c| c == ',' || c == '\r' || c == '\n' || c == ' ')
+        {
+            None
+        } else {
+            Some(Tag(s))
+        }
+    }
+
+    /// Returns the tag contents as a str.
+    pub fn as_str(&self) -> &'a str {
+        self.0
+    }
+}
+
+impl<'a> ops::Deref for Tag<'a> {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        self.as_str()
+    }
+}
+
+/// An individual interval histogram parsed from an interval log.
+#[derive(PartialEq, Debug)]
+pub struct IntervalLogHistogram<'a> {
+    tag: Option<Tag<'a>>,
+    start_timestamp: f64,
+    // lazily map to Duration to save parsing time and a few bytes of space on ILH
+    duration: f64,
+    max: f64,
+    encoded_histogram: &'a str,
+}
+
+impl<'a> IntervalLogHistogram<'a> {
+    /// Tag, if any is present.
+    pub fn tag(&self) -> Option<Tag<'a>> {
+        self.tag
+    }
+
+    /// Timestamp of the start of the interval in seconds.
+    ///
+    /// The timestamp may be absolute vs the epoch, or there may be a `StartTime` or `BaseTime` for
+    /// the log, in which case you may wish to consider this number as a delta vs those timestamps.
+    /// See the module-level documentation about timestamps.
+    pub fn start_timestamp(&self) -> f64 {
+        self.start_timestamp
+    }
+
+    /// Duration of the interval in seconds.
+    pub fn duration(&self) -> time::Duration {
+        let secs = self.duration as u64;
+        // can't overflow because this can be at most 1 billion which is approx 2^30
+        let nsecs = (self.duration.fract() * 1_000_000_000_f64) as u32;
+        time::Duration::new(secs, nsecs)
+    }
+
+    /// Max value in the encoded histogram
+    ///
+    /// This max value is the max of the histogram divided by some scaling factor (which may be
+    /// 1.0).
+    pub fn max(&self) -> f64 {
+        self.max
+    }
+
+    /// Base64-encoded serialized histogram.
+    ///
+    /// If you need the deserialized histogram, base64-decode and use a `Deserializer` on the
+    /// resulting bytes.
+    pub fn encoded_histogram(&self) -> &'a str {
+        self.encoded_histogram
+    }
+}
+
+#[derive(PartialEq, Debug)]
+/// Represents one non-comment line in an interval log.
+///
+/// One thing to note is that the way your interval timestamps work can vary. If your log was
+/// written with a StartTime or BaseTime, that metadata will appear in header comments, and that
+/// will be represented by the iterator providing the corresponding variants here. The presence
+/// of those timestamps will affect how you should interpret the timestamps for individual
+/// intervals. See the module-level documentation.
+#[allow(variant_size_differences)]
```

Comment:
Should we care about this? Not really sure I see a way around it

Reply:
I thought `variant_size_differences` was allowed by default? I think it's fine for this `enum` to have a somewhat large size variant, though it is also another reason to make the iterator only yield `IntervalLogHistogram` (as the enum would no longer be necessary).

Reply:
They are, but they're turned off in the crate-wide `deny` statement. The variant is only large because `IntervalLogHistogram` is large, though, so removing the enum isn't really going to help the "copying fat structs around" issue (though I think it's unlikely to be a performance issue).